File: Network/UpdateGeph.py
#!/usr/bin/env python3

from csv import DictReader
from io import BytesIO
from subprocess import Popen, check_call, check_output
from time import time

import requests
from click import pause

proxies = {'https': 'http://localhost:1080'}

api = 'https://api.github.com/repos/geph-official/gephng-binaries/commits'


def getDownloadURL():
    sha = requests.get(api, proxies=proxies, timeout=10).json()[0]['sha']
    files = [f['raw_url']
             for f in requests.get(api+'/'+sha, proxies=proxies, timeout=10).json()['files']]
    for f in files:
        if 'geph-client-windows' in f:
            target = f
            return target


def checkVersion(new):
    try:
        with open('version', 'r') as f:
            current = f.read().strip()
    except FileNotFoundError:
        print('Check version failed.')
        return 1
    if current == new:
        print(f'Latest version v{new} is already installed.')
        return 0
    else:
        print('New version found.')
        return 1


def download(target):
    start = time()
    response = requests.get(target, proxies=proxies, timeout=10, stream=True)
    total = int(response.headers['content-length'])
    downloaded = 0
    content = BytesIO()
    print(f"[{' '*50}]   0.0 KB/s -   0.0% of {total/1048576:.2f}M", end='\r')
    for chunk in response.iter_content(1024*512):
        if chunk:
            downloaded += len(chunk)
            content.write(chunk)
            done = int(50 * downloaded / total)
            speed = downloaded/(time()-start)/1024
            left = (total-downloaded)/(speed*1024)
            if speed >= 1024:
                speed = f'{speed/1024:.1f} MB/s'
            else:
                speed = f'{speed:.1f} KB/s'
            if left >= 60:
                left = f'{left/60:.1f} mins left'
            else:
                left = f'{left:.0f} secs left'
            print(
                f"[{'='*done}{' '*(50-done)}] {speed:>10} - {downloaded/total:>6.1%} of {total/1048576:.2f}M, {left:<14}", end='\r')
    if downloaded != total:
        raise Warning('Download failed.')
    print(f"[{'='*done}{' '*(50-done)}] {speed:>10} - {downloaded/total:>6.1%} of {total/1048576:.2f}M, {left:<14}")
    print(f'Download finished. Total time: {time()-start:.2f}s')
    running = shutdown()
    try:
        with open('geph.exe', 'wb') as f:
            f.write(content.getvalue())
    except PermissionError:
        with open('geph.tmp', 'wb') as f:
            f.write(content.getvalue())
        raise Warning('Failed to update geph.exe file.')
    return running


def shutdown():
    tasks = DictReader(check_output('tasklist /fo csv').decode().splitlines())
    for t in tasks:
        if 'geph' in t['Image Name']:
            print('Shutting down Geph service.')
            check_call('taskkill /f /pid '+t['PID'])
            return 1
    return 0


def main():
    try:
        target = getDownloadURL()
    except:
        print('Failed to get download URL.')
        return
    tag_name = target.split('-v')[1].replace('.exe', '')
    if checkVersion(tag_name):
        print(f'Downloading v{tag_name}')
        try:
            running = download(target)
        except Exception as e:
            print(str(e))
            return
        with open('version', 'w') as f:
            f.write(tag_name)
        if running:
            print('Restarting Geph service.')
            Popen('geph.exe -config client.conf',
                  creationflags=134217728)  # CREATE_NO_WINDOW


if __name__ == '__main__':
    main()
    pause()

File: windows/UpdateGeph.go
package main

import (
	"bytes"
	"fmt"
	"io"
	"io/ioutil"
	"log"
	"net/http"
	"os"
	"os/exec"
	"strconv"
	"strings"
	"time"

	ps "github.com/mitchellh/go-ps"
	"github.com/sunshineplan/gohttp"
	"github.com/sunshineplan/utils/progressbar"
)

const api = "https://api.github.com/repos/geph-official/gephng-binaries/commits"

var client = http.DefaultClient

var tag, downloadURL string
var process bool

var pb *progressbar.ProgressBar

type writeCounter struct{}

func (wc *writeCounter) Write(p []byte) (int, error) {
	pb.Add(len(p))
	return 0, nil
}

func main() {
	getDownloadURL()
	if checkVersion() {
		fmt.Printf("Downloading v%s\n", tag)
		b := download()
		checkProcess()
		replace(b)
		version, err := os.Create("version")
		if err != nil {
			log.Fatal(err)
		}
		defer version.Close()
		if _, err := version.WriteString(tag); err != nil {
			log.Fatal(err)
		}
		if process {
			fmt.Println("Restarting Geph service.")
			if err := exec.Command("python", "-c",
				"import subprocess;subprocess.Popen('geph -config client.conf',creationflags=134217728)").Run(); err != nil {
				log.Fatalln("Failed to restart geph:", err)
			}
		}
	}
	fmt.Println("Press enter key to continue . . .")
	fmt.Scanln()
}

func getDownloadURL() {
	var commits []struct {
		SHA string
	}
	if err := gohttp.GetWithClient(api, nil, client).JSON(&commits); err != nil {
		log.Fatal(err)
	}
	var commit struct {
		Files []struct {
			RawURL string `json:"raw_url"`
		}
	}
	if err := gohttp.GetWithClient(api+"/"+commits[0].SHA, nil, client).JSON(&commit); err != nil {
		log.Fatal(err)
	}
	for _, i := range commit.Files {
		if strings.Contains(i.RawURL, "geph-client-windows") {
			tag = strings.ReplaceAll(strings.Split(i.RawURL, "-v")[1], ".exe", "")
			downloadURL = i.RawURL
			return
		}
	}
	log.Fatal("Not found")
}

func checkVersion() bool {
	current, err := ioutil.ReadFile("version")
	if err != nil {
		fmt.Println("Check version failed.")
		return true
	}
	result := tag != strings.TrimSpace(string(current))
	if result {
		fmt.Println("New version found.")
	} else {
		fmt.Printf("Latest version v%s is already installed.\n", tag)
	}
	return result
}

func download() []byte {
	resp := gohttp.GetWithClient(downloadURL, nil, client)
	if resp.Error != nil {
		log.Fatal(resp.Error)
	}
	total, err := strconv.Atoi(resp.Header.Get("content-length"))
	if err != nil {
		log.Fatal(err)
	}
	c := make(chan bool, 1)
	pb = progressbar.New(total, c)
	pb.Start()
	var out bytes.Buffer
	if _, err := io.Copy(&out, io.TeeReader(resp.Body, &writeCounter{})); err != nil {
		log.Fatal(err)
	}
	defer resp.Body.Close()
	<-c
	return out.Bytes()
}

func checkProcess() {
	processes, err := ps.Processes()
	if err != nil {
		log.Fatal(err)
	}
	for _, p := range processes {
		if strings.Contains(p.Executable(), "geph") {
			fmt.Println("Shutting down Geph service.")
			process = true
			if err := (&os.Process{Pid: p.Pid()}).Kill(); err != nil {
				log.Fatal(err)
			}
			time.Sleep(time.Second)
			break
		}
	}
}

func replace(b []byte) {
	f, err := os.Create("geph.exe")
	if err != nil {
		f, err = os.Create("geph.tmp")
		if err != nil {
			log.Fatal(err)
		} else {
			defer log.Fatal("Failed to replace geph.exe file.")
		}
	}
	defer f.Close()
	if _, err := f.Write(b); err != nil {
		log.Fatal(err)
	}
}
